atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,294 @@
+        Set<String> hostWithEndpoint = Set.of(newHostOne);
+        setHostIds(
+                filterContainers(allHosts, server -> !hostWithEndpoint.contains(server)), HostIdResult.softFailure());
+        setHostIds(filterContainers(allHosts, hostWithEndpoint::contains), DEFAULT_RESULT);
+        assertThat(validator.getNewHostsWithInconsistentTopologies(allHosts.keySet(), allHosts))
+                .isEmpty();
+    }
+
+    @Test
+    public void validateNewlyAddedHostsAddsAllHostsIfEndpointExistingHostIdsMatch() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<String> hostsWithEndpoints = ImmutableSet.of(newHostOne, oldHostOne);
+        setHostIds(
+                filterContainers(allHosts, server -> !hostsWithEndpoints.contains(server)), HostIdResult.softFailure());
+        setHostIds(filterContainers(allHosts, hostsWithEndpoints::contains), DEFAULT_RESULT);
+        assertThat(validator.getNewHostsWithInconsistentTopologies(
+                        filterServers(allHosts, newHosts::contains), allHosts))
+                .isEmpty();
+    }
+
+    @Test
+    public void validateNewlyAddedHostsReturnsNewHostsUnlessEndpointExistAndDoNotMatch() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<String> hostsWithEndpoints = ImmutableSet.of(newHostOne, oldHostOne);
+        setHostIds(
+                filterContainers(allHosts, server -> !hostsWithEndpoints.contains(server)), HostIdResult.softFailure());
+        setHostIds(filterContainers(allHosts, newHostOne::equalsIgnoreCase), HostIdResult.success(Set.of("uuid")));
+        setHostIds(filterContainers(allHosts, oldHostOne::equalsIgnoreCase), DEFAULT_RESULT);
+        assertThat(validator.getNewHostsWithInconsistentTopologies(
+                        filterServers(allHosts, newHosts::contains), allHosts))
+                .containsExactlyElementsOf(filterServers(allHosts, newHostOne::equalsIgnoreCase));
+    }
+
+    @Test
+    public void validateNewlyAddedHostsNoNewHostsAddedIfOldHostsDoNotHaveQuorum() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<CassandraServer> newCassandraServers = filterServers(allHosts, newHosts::contains);
+        Set<String> hostsOffline = ImmutableSet.of(oldHostOne, oldHostTwo);
+        setHostIds(filterContainers(allHosts, hostsOffline::contains), HostIdResult.hardFailure());
+        setHostIds(filterContainers(allHosts, server -> !hostsOffline.contains(server)), HostIdResult.success(uuids));
+        assertThat(validator.getNewHostsWithInconsistentTopologies(newCassandraServers, allHosts))
+                .containsExactlyElementsOf(newCassandraServers);
+    }
+
+    @Test
+    public void validateNewlyAddedHostsNoNewHostsAddedIfNewHostsDoNotHaveQuorum() {
```

Comment:
This is only when there are no current servers right? The test name should reflect that